### PR TITLE
Add keep transformer including tests

### LINKED ISSUE #3: Add `keep` dropper
> I find very annoying to `drop` every column I temporarly create. It would be a nice feature to specify the columns to keep instead of dropping the useless ones.
> 
> E.g. I have columns `a,b,c,d,e,f,g` and want to keep only `a,b`:
> 
> ```
> // current
> stream | drop<c,d,e,f,g>
> 
> // proposed
> stream | keep<a,b>
> ```
> This would be useful in the end of the streams on in case of reusing some stream parts (no idea what sources are provided)

### PR DESCRIPTION
**Pull request by blazekadam**
closes #3

### CODE REVIEW

File: include/hipipe/core/stream/keep.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/****************************************************************************
+ *  hipipe library
+ *  Copyright (c) 2017, Cognexa Solutions s.r.o.
+ *  Copyright (c) 2018, Iterait a.s.
+ *  Author(s) Filip Matzner, Adam Blažek
```
Comment (FloopCZ) [inline include/hipipe/core/stream/keep.hpp]:
```suggestion
 *  Author(s) Adam Blažek
```

File: include/hipipe/core/stream/keep.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+/// \ingroup Stream
+/// \brief Keep the specified columns in the stream, drop everything else.
+///
+/// Example:
+/// \code
+///     HIPIPE_DEFINE_COLUMN(id, int)
+///     HIPIPE_DEFINE_COLUMN(value, double)
+///     std::vector<std::tuple<int, double>> data = {{3, 5.}, {1, 2.}};
+///     auto rng = data | create<id, value>() | keep<value>; // now it has only the value column
```
Comment (FloopCZ) [inline include/hipipe/core/stream/keep.hpp]:
```suggestion
///     auto rng = data | create<id, value>() | keep<value>;  // now it has only the value column
```

File: test/core/stream/CMakeLists.txt [hunk 3]
```diff
@@ -8,6 +8,8 @@ add_boost_test("test.core.stream.create" "create.cpp" "")
 
 add_boost_test("test.core.stream.drop" "drop.cpp" "")
 
+add_boost_test("test.core.stream.keep" "keep.cpp" "")
```
Comment (FloopCZ) [inline test/core/stream/CMakeLists.txt]: Please sort alphabetically.

File: test/core/stream/keep.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/****************************************************************************
+ *  hipipe library
+ *  Copyright (c) 2017, Cognexa Solutions s.r.o.
```
Comment (FloopCZ) [inline test/core/stream/keep.cpp]:
```suggestion
```

File: test/core/stream/keep.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *  Copyright (c) 2018, Iterait a.s.
+ *  Author(s) Filip Matzner, Adam Blažek
```
Comment (FloopCZ) [inline test/core/stream/keep.cpp]:
```suggestion
 *  Author(s) Adam Blažek
```

File: test/core/stream/keep.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE column_keep_test
+
+#include "common.hpp"
+
+#include <hipipe/core/stream/keep.hpp>
+
+HIPIPE_DEFINE_COLUMN(Unique3, std::unique_ptr<int>)
+
+
```
Comment (FloopCZ) [inline test/core/stream/keep.cpp:18]:
```suggestion
```

File: include/hipipe/core/stream/keep.hpp [hunk 7]
```diff
@@ -0,0 +1,70 @@
+/****************************************************************************
+ *  hipipe library
+ *  Copyright (c) 2017, Cognexa Solutions s.r.o.
```
Comment (FloopCZ) [inline include/hipipe/core/stream/keep.hpp]:
```suggestion
```

File: include/hipipe/core/stream/batch_t.hpp [hunk 8]
```diff
@@ -111,6 +132,27 @@ class batch {
           std::make_unique<Column>(std::forward<Args>(args)...));
     }
 
+    /// \brief Insert a raw column handle to the batch or rewrite an existing one.
+    ///
+    /// This is particularly useful when a raw column handle is retrieved with the at() function and there is no need
```
Comment (FloopCZ) [inline include/hipipe/core/stream/batch_t.hpp]: Please wrap this line.

File: include/hipipe/core/stream/batch_t.hpp [hunk 9]
```diff
@@ -111,6 +132,27 @@ class batch {
...
+    /// for creating a new column.
```
Comment (FloopCZ) [inline include/hipipe/core/stream/batch_t.hpp]: I like the comment, thanks.

File: test/core/stream/batch_t.cpp [hunk 10]
```diff
@@ -29,6 +29,18 @@ BOOST_AUTO_TEST_CASE(test_batch_insert_assign_extract)
...
+    using hipipe::stream::batch_t;
+
+    batch_t batch, batch2;
+    batch.insert_or_assign<Int>(std::vector<int>{0, 1, 2});
+    BOOST_TEST(batch.at<Int>()->extract<Int>() == std::vector<int>({0, 1, 2}));
+    BOOST_CHECK_THROW(batch.at<Double>(), std::runtime_error);
+
+    batch2.raw_insert_or_assign<Int>(std::move(batch.at<Int>()));
+    BOOST_TEST(batch2.at<Int>()->extract<Int>() == std::vector<int>({0, 1, 2}));
+}
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp:45]:
```suggestion


```

File: test/core/stream/batch_t.cpp [hunk 11]
```diff
@@ -29,6 +29,18 @@ BOOST_AUTO_TEST_CASE(test_batch_insert_assign_extract)
     BOOST_CHECK_THROW(batch.extract<Double>(), std::runtime_error);
 }
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp:31]:
```suggestion


```

File: test/core/stream/keep.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ ****************************************************************************/
+
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE column_keep_test
+
+#include "common.hpp"
+
+#include <hipipe/core/stream/keep.hpp>
+
+HIPIPE_DEFINE_COLUMN(Unique3, std::unique_ptr<int>)
```
Comment (FloopCZ) [inline test/core/stream/keep.cpp]: Why not decrease the number to 2? :-D

File: test/core/stream/batch_t.cpp [hunk 13]
```diff
@@ -29,6 +29,18 @@ BOOST_AUTO_TEST_CASE(test_batch_insert_assign_extract)
...
+    using hipipe::stream::batch_t;
+
+    batch_t batch, batch2;
+    batch.insert_or_assign<Int>(std::vector<int>{0, 1, 2});
+    BOOST_TEST(batch.at<Int>()->extract<Int>() == std::vector<int>({0, 1, 2}));
+    BOOST_CHECK_THROW(batch.at<Double>(), std::runtime_error);
+
+    batch2.raw_insert_or_assign<Int>(std::move(batch.at<Int>()));
+    BOOST_TEST(batch2.at<Int>()->extract<Int>() == std::vector<int>({0, 1, 2}));
+}
```
Comment (FloopCZ) [inline test/core/stream/batch_t.cpp:44]: I have to admit I would prefer to test those functions separately as everywhere in this project where it is possible. But not that much to let you rewrite it.

**Review by FloopCZ — COMMENTED** [review]
Thank you for the changes! This is better.

File: test/core/stream/batch_t.cpp (hunk 13 above)
Comment (blazekadam) [reply test/core/stream/batch_t.cpp:44]: I feel you.

**Review by FloopCZ — APPROVED** [review]
Thank you @blazekadam for your code! And welcome to the contributor party! :wink:

**Pull request merged**